manager/allocator/network/port/port.go

```diff
@@ -0,0 +1,470 @@
+// reassigning the same dynamically allocated port number that is being removed
+// to a new port that is being added. However, this behavior _is not tested_
+// and should _not_ be relied on as part of the API. It's a nice-to-have
+// implementation detail.
+//
+// Because I'm reluctant to alter what I have that already works, we will not
+// return ErrAlreadyAllocated if the endpoint is already allocated. Instead,
+// callers can use the AlreadyAllocated function from this package.
+func (pa *allocator) Allocate(endpoint *api.Endpoint, spec *api.EndpointSpec) (Proposal, error) {
+	// Ok, so Allocate is actually pretty tricky, because we do dynamic
+	// port allocation. This means if the user gives us no published port, we
+	// pick one for them. When we're updating ports, we have to figure out if
+	// the user has *not* updated a dynamically assigned port, so we can use
+	// the same port number after the update.
+	//
+	// The naive solution is to just check if the object has a published port
+	// for a given target port, but the spec does not. Then we could just reuse
+	// the port on the object, right? Except if the user goes from providing a
+	// published port to dynamically assigning a published port. In that case,
+	// the actual state of the object will have a published port, but the
+	// user's spec wants a new (dynamically allocated) port, just like if we
+	// had dynamically assigned the port.
+	//
+	// Luckily for use, we keep a copy of the spec on the Endpoint object,
+	// which we own, which means we have the old endpoint spec around and can
+	// compare the user's specs. Then, all we have to do is see if the
+	// published port changed.
+	if spec == nil {
+		spec = &api.EndpointSpec{}
+	}
+
+	// So, basically, here's what we're going to do: we're going to create a
+	// new list of PortConfigs. This will be the final list that gets put into
+	// the object endpoint object
+	finalPorts := make([]*api.PortConfig, len(spec.Ports))
+
+	// first, we need to "recover" any dynamically assigned publish ports. to
+	// do this:
+	//   1. go through all of the ports in the new spec.
+	//   2. if a port doesn't have a publish port assigned
+	//     a. go through all of the ports in the old, checking if one matches
+	//     b. if so, that port has a dynamic port assignment
+	//        i. go through every port in the old object's ports. if every
+	//           component of the port is the same EXCEPT the PublishedPort,
+	//           copy that published port into the port assignments.
+	for i, spec := range spec.Ports {
+		// check if the published port or target port is off the end of the
+		// allowed port range
+		if spec.PublishedPort > masterPortEnd {
+			return nil, errors.ErrInvalidSpec("published port %v isn't in the valid port range", spec.PublishedPort)
+		}
+		if spec.TargetPort > masterPortEnd {
+			return nil, errors.ErrInvalidSpec("target port %v isn't in the valid port range", spec.TargetPort)
+		}
```

Comment:
This probably needs a check for `masterPortStart` too. For `PublishedPort`, it's not actually a problem (unless someone changes the constant) because the only value less that 1 means dynamic allocation. But I guess we shouldn't allow `TargetPort` to be 0?